src/main/com/kubelt/lib/wallet/node.cljs

```
@@ -64,111 +73,120 @@
 ;; TODO allow wallet listing
 ;; TODO allow wallet deletion
 
-(defn has-wallet?
+(defn has-wallet?&
   "Return true if the named wallet already exists."
   [app-name wallet-name]
-  (let [wallet-path (name->path app-name wallet-name)]
-    (.existsSync fs wallet-path)))
+  (fs-exists?& (name->path app-name wallet-name)))
 
-(defn can-decrypt?
+(defn can-decrypt?&
   "Return true if the wallet can be successfully decrypted with the
   supplied password, and false otherwise."
   [app-name wallet-name password]
-  (if (has-wallet? app-name wallet-name)
-    (let [wallet-path (name->path app-name wallet-name)
-          wallet-str (.readFileSync fs wallet-path)]
-      (try
-        (.fromEncryptedJsonSync Wallet wallet-str password)
-        true
-        (catch js/Error e
-          ;; Error: invalid password
-          false)))
-    false))
+  (-> (has-wallet?& app-name wallet-name)
+      (lib.promise/then
+       (fn [wallet-path]
+         (->
+          (.readFile fs-promises wallet-path)
+          (lib.promise/then #(.fromEncryptedJson Wallet % password))
+          (lib.promise/then (fn [_] true))
+          (lib.promise/catch (fn [e]
```

Should we avoid catching an error here, leaving it to the caller to supply exception handling behavior? If we mask errors that occur here, will the caller have another path to discovering what went wrong, e.g. supplied an invalid path to a wallet, rather than an invalid password that prevents successful decryption. 🤔

---

after a few more work on how promises work and how we did, I think code is better now